<h1>Lester Genton </h1>

<h1>BSIT 3A</h1>

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
accident = pd.read_csv('datasets//uk_road_accident.csv')


In [3]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [5]:
accident.describe().T

count       mean       std        min        25%  \
Latitude              660654.0  52.553866  1.406922  49.914430  51.490690   
Longitude             660653.0  -1.431210  1.383330  -7.516225  -2.332291   
Number_of_Casualties  660679.0   1.357040  0.824847   1.000000   1.000000   
Number_of_Vehicles    660679.0   1.831255  0.715269   1.000000   1.000000   

                            50%        75%        max  
Latitude              52.315641  53.453452  60.757544  
Longitude             -1.411667  -0.232869   1.762010  
Number_of_Casualties   1.000000   1.000000  68.000000  
Number_of_Vehicles     2.000000   2.000000  32.000000

In [6]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [7]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode()[0])


In [8]:
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode()[0])


In [9]:
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna('unknown condition')


In [10]:
accident['Road_Type'] = accident['Road_Type'].fillna(accident['Road_Type'].mode()[0])


In [11]:
accident['Urban_or_Rural_Area'] 

0         Urban
1         Urban
2         Urban
3         Urban
4         Urban
          ...  
660674    Rural
660675    Rural
660676    Rural
660677    Rural
660678    Rural
Name: Urban_or_Rural_Area, Length: 660679, dtype: object

In [12]:
accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])


0         Urban
1         Urban
2         Urban
3         Urban
4         Urban
          ...  
660674    Rural
660675    Rural
660676    Rural
660677    Rural
660678    Rural
Name: Urban_or_Rural_Area, Length: 660679, dtype: object

In [13]:
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna('unknown weather')


In [14]:
accident.isnull().sum()


Index                       0
Accident_Severity           0
Accident Date               0
Latitude                    0
Light_Conditions            0
District Area               0
Longitude                   0
Number_of_Casualties        0
Number_of_Vehicles          0
Road_Surface_Conditions     0
Road_Type                   0
Urban_or_Rural_Area        15
Weather_Conditions          0
Vehicle_Type                0
dtype: int64

In [15]:
accident.dtypes


Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [16]:
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')

In [17]:
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')


In [18]:
accident['Latitude'] = accident['Latitude'].astype('category')


In [19]:
accident['District Area'] = accident['District Area'].astype('category')


In [20]:
accident['Longitude'] = accident['Longitude'].astype('category')

In [21]:
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')

In [22]:
accident['Road_Type'] = accident['Road_Type'].astype('category')


In [23]:
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')

In [24]:
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')

In [25]:
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')


In [26]:
accident.isnull().sum()


Index                       0
Accident_Severity           0
Accident Date               0
Latitude                    0
Light_Conditions            0
District Area               0
Longitude                   0
Number_of_Casualties        0
Number_of_Vehicles          0
Road_Surface_Conditions     0
Road_Type                   0
Urban_or_Rural_Area        15
Weather_Conditions          0
Vehicle_Type                0
dtype: int64

In [27]:
accident['Accident Date'] = accident['Accident Date'].astype('str')
accident['Accident Date'] = accident['Accident Date'].str.strip()
accident['Accident Date'] = accident['Accident Date'].str.replace('/','-')
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'],dayfirst=True,errors = 'coerce')


In [28]:
accident.isnull().sum()

Index                       0
Accident_Severity           0
Accident Date               0
Latitude                    0
Light_Conditions            0
District Area               0
Longitude                   0
Number_of_Casualties        0
Number_of_Vehicles          0
Road_Surface_Conditions     0
Road_Type                   0
Urban_or_Rural_Area        15
Weather_Conditions          0
Vehicle_Type                0
dtype: int64

In [29]:
accident['Year'] = accident['Accident Date'].dt.year
accident['Month'] = accident['Accident Date'].dt.month
accident['DayofWeek'] = accident['Accident Date'].dt.dayofweek
accident['Day'] = accident['Accident Date'].dt.day

In [30]:
accident.isnull().sum()

Index                       0
Accident_Severity           0
Accident Date               0
Latitude                    0
Light_Conditions            0
District Area               0
Longitude                   0
Number_of_Casualties        0
Number_of_Vehicles          0
Road_Surface_Conditions     0
Road_Type                   0
Urban_or_Rural_Area        15
Weather_Conditions          0
Vehicle_Type                0
Year                        0
Month                       0
DayofWeek                   0
Day                         0
dtype: int64

In [31]:
accident['Year'] = accident['Year'].astype('category')
accident['Month'] = accident['Month'].astype('category')
accident['DayofWeek'] = accident['DayofWeek'].astype('category')
accident['Day'] = accident['Day'].astype('category')

In [32]:
accident.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                         category
Light_Conditions                 category
District Area                    category
Longitude                        category
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
Year                             category
Month                            category
DayofWeek                        category
Day                              category
dtype: object

<h1>Exploratory Data Analytics</h1>

<h1>1. Which area has many accidents?</h1>

In [33]:
accident['District Area'].unique()

['Kensington and Chelsea', 'Westminster', 'Richmond upon Thames', 'Hammersmith and Fulham', 'Hounslow', ..., 'County Durham', 'Shropshire', 'Central Bedfordshire', 'Cornwall', 'Wiltshire']
Length: 422
Categories (422, object): ['Aberdeen City', 'Aberdeenshire', 'Adur', 'Allerdale', ..., 'Wycombe', 'Wyre', 'Wyre Forest', 'York']

In [34]:
accident[accident['District Area'] == 'Kensington and Chelsea'].value_counts().sum()

np.int64(2799)

In [35]:
accident[accident['District Area'] == 'Westminster'].value_counts().sum()

np.int64(5706)

In [36]:
accident[accident['District Area'] == 'Richmond upon Thames'].value_counts().sum()

np.int64(1607)

In [37]:
accident[accident['District Area'] == 'Hammersmith and Fulham'].value_counts().sum()

np.int64(2509)

In [38]:
accident[accident['District Area'] == 'Hounslow'].value_counts().sum()

np.int64(2979)

In [39]:
accident[accident['District Area'] == '...'].value_counts().sum()

np.int64(0)

In [40]:
accident[accident['District Area'] == 'County Durham'].value_counts().sum()

np.int64(2228)

In [41]:
accident[accident['District Area'] == 'Shropshire'].value_counts().sum()

np.int64(1343)

In [42]:
accident[accident['District Area'] == 'Central Bedfordshire'].value_counts().sum()

np.int64(1225)

In [43]:
accident[accident['District Area'] == 'Cornwall'].value_counts().sum()

np.int64(2606)

In [44]:
accident[accident['District Area'] == 'Wiltshire'].value_counts().sum()

np.int64(1590)

<h1>Insight 1</h1>

<h2>Based on all areas there are many accidents, but the area of westminster has the highest counts of accident.</h2>

<h1>2. How many incidents have 5 or more casualities that include?</h1>

In [45]:
accident[accident['Number_of_Casualties'] > 5].value_counts().sum()

np.int64(2612)

<h1>Insight 2.</h1>

<h2>There have been 2612 accidents have 5 or more casualities in each incident.</h2>

<h1>3. What are the common and most frequent accidents associated with road types?</h1>

In [46]:
accident['Road_Type'].unique()

['Single carriageway', 'Dual carriageway', 'One way street', 'Roundabout', 'Slip road']
Categories (5, object): ['Dual carriageway', 'One way street', 'Roundabout', 'Single carriageway', 'Slip road']

In [47]:
accident[accident['Road_Type'] == 'Single carriageway'].value_counts().sum()

np.int64(496653)

In [48]:
accident[accident['Road_Type'] == 'Dual carriageway'].value_counts().sum()

np.int64(99420)

In [49]:
accident[accident['Road_Type'] == 'One way street'].value_counts().sum()

np.int64(13558)

In [50]:
accident[accident['Road_Type'] == 'Roundabout'].value_counts().sum()

np.int64(43992)

In [51]:
accident[accident['Road_Type'] == 'Slip road'].value_counts().sum()

np.int64(7041)

<h1>Insights 3.</h1>
<h2>The common road had more accidents. It was a Single carriageway, because somehow, many cars were traveling at high speeds.</h2>

<h1>4. Get all accidents in the Rural Area that have 1 casualty only</h1>


In [52]:
accident[(accident['Urban_or_Rural_Area'] == 'Rural') & (accident['Number_of_Casualties'])]

Index Accident_Severity Accident Date   Latitude  \
6101    200701JC30253            Slight    2019-04-29  51.616921   
6193    200701JC63842            Slight    2019-06-23  51.634035   
6223    200701JC64859            Slight    2019-07-15  51.616088   
6421    200701JC67563            Slight    2019-10-17  51.634365   
6441    200701JC67754            Slight    2019-11-03  51.615152   
...               ...               ...           ...        ...   
660672  201091NM01085           Serious    2022-01-29  57.170725   
660675  201091NM01881            Slight    2022-02-21  57.232273   
660676  201091NM01935            Slight    2022-02-23  57.585044   
660677  201091NM01964           Serious    2022-02-23  57.214898   
660678  201091NM02142           Serious    2022-02-28  57.575210   

                   Light_Conditions   District Area Longitude  \
6101          Darkness - lights lit  Waltham Forest  0.010797   
6193                       Daylight  Waltham Forest  0.014444   
6223    Darkness - lighting unknown  Waltham Forest  0.012204   
6421         Darkness - no lighting  Waltham Forest  0.021395   
6441                       Daylight  Waltham Forest  0.014330   
...                             ...             ...       ...   
660672                     Daylight        Highland -3.852421   
660675       Darkness - no lighting        Highland -3.809281   
660676                     Daylight        Highland -3.862727   
660677       Darkness - no lighting        Highland -3.823997   
660678                     Daylight        Highland -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
6101                       1                   1                     Dry   
6193                       1                   2             Wet or damp   
6223                       1                   3                     Dry   
6421                       1                   1             Wet or damp   
6441                       1                   2                     Dry   
...                      ...                 ...                     ...   
660672                     1                   2                    Snow   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
6101    Single carriageway               Rural     Fine no high winds   
6193    Single carriageway               Rural  Raining no high winds   
6223    Single carriageway               Rural     Fine no high winds   
6421    Single carriageway               Rural  Raining no high winds   
6441    Single carriageway               Rural     Fine no high winds   
...                    ...                 ...                    ...   
660672  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                 Vehicle_Type  Year Month DayofWeek Day  
6101                      Car  2019     4         0  29  
6193    Motorcycle over 500cc  2019     6         6  23  
6223                      Car  2019     7         0  15  
6421                      Car  2019    10         3  17  
6441                      Car  2019    11         6   3  
...                       ...   ...   ...       ...  ..  
660672                    Car  2022     1         5  29  
660675                    Car  2022     2         0  21  
660676                    Car  2022     2         2  23  
660677  Motorcycle over 500cc  2022     2       

<h1>Insights 4.</h1>
<h2>We can see above all the accidents in the Rural area.</h2>

<h1> 5. Which incident has the highest Vehicles involved?</h1>

In [53]:
accident['Number_of_Vehicles'].unique() 

array([ 2,  3,  4,  1,  5,  6, 14,  8,  7, 11, 10,  9, 16, 13, 12, 28, 15,
       32, 19])

In [54]:
accident[accident['Number_of_Vehicles'] == 32]

Index Accident_Severity Accident Date   Latitude  \
497581  2009559D05333            Slight    2021-12-23  50.752096   

       Light_Conditions District Area Longitude  Number_of_Casualties  \
497581         Daylight       Purbeck -2.254705                     5   

        Number_of_Vehicles Road_Surface_Conditions         Road_Type  \
497581                  32            Frost or ice  Dual carriageway   

       Urban_or_Rural_Area  Weather_Conditions Vehicle_Type  Year Month  \
497581               Rural  Fine no high winds          Car  2021    12   

       DayofWeek Day  
497581         3  23

<h1>Insights</h1>
<h2>The Vehicle involved in the incident is 32 Vehicles.</h2>

<h1>6. Retrieve all Car incidents that have involved three or more additional vehicles.</h1>

In [55]:
accident[(accident['Vehicle_Type'] == 'Car') & (accident['Number_of_Vehicles'] > 2)]

Index Accident_Severity Accident Date   Latitude  \
5       200701BS67159           Serious    2019-09-18  51.497750   
19      200701BS68237           Serious    2019-11-26  51.482988   
248     200701BS70255           Serious    2019-05-16  51.491632   
336     200701BS70350            Slight    2019-06-22  51.485178   
354     200701BS70368            Slight    2019-06-29  51.482076   
...               ...               ...           ...        ...   
660620  201091NK04948            Slight    2022-10-04  59.047179   
660641  201091NL04969            Slight    2022-09-04  60.151776   
660666  201091NM00770            Slight    2022-01-21  56.879446   
660668  201091NM00839            Slight    2022-01-23  57.054889   
660676  201091NM01935            Slight    2022-02-23  57.585044   

                   Light_Conditions           District Area Longitude  \
5                          Daylight  Kensington and Chelsea -0.192556   
19                         Daylight  Kensington and Chelsea -0.174272   
248                        Daylight  Kensington and Chelsea -0.192366   
336                        Daylight  Kensington and Chelsea -0.187868   
354           Darkness - lights lit  Kensington and Chelsea -0.173445   
...                             ...                     ...       ...   
660620                     Daylight          Orkney Islands -3.196707   
660641        Darkness - lights lit        Shetland Islands -1.148217   
660666  Darkness - lighting unknown                Highland -4.255357   
660668                     Daylight                Highland -4.130434   
660676                     Daylight                Highland -3.862727   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
5                          2                   3                     Dry   
19                         1                   4                     Dry   
248                        1                   3                     Dry   
336                        1                   3                     Dry   
354                        1                   3             Wet or damp   
...                      ...                 ...                     ...   
660620                     6                   3                     Dry   
660641                     1                   3                     Dry   
660666                     2                   4                    Snow   
660668                     2                   3            Frost or ice   
660676                     1                   3            Frost or ice   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
5       Single carriageway               Urban     Fine no high winds   
19      Single carriageway               Urban     Fine no high winds   
248     Single carriageway               Urban     Fine no high winds   
336         One way street               Urban     Fine no high winds   
354     Single carriageway               Urban  Raining no high winds   
...                    ...                 ...                    ...   
660620          Roundabout               Rural     Fine no high winds   
660641  Single carriageway               Rural     Fine no high winds   
660666  Single carriageway               Rural   Snowing + high winds   
660668  Single carriageway               Rural                  Other   
660676  Single carriageway               Rural     Fine no high winds   

       Vehicle_Type  Year Month DayofWeek Day  
5               Car  2019     9         2  18  
19              Car  2019    11         1  26  
248             Car  2019     5         3  16  
336             Car  2019     6         5  22  
354             Car  2019     6         5  29  
...             ...   ...   ...       ...  ..  
660620          Car  2022    10         1   4  
660641          Car  2022     9         6   4  
660666          Car  2022     1         4  21  
660668          Car  2022     1         6  23  
6606

<h1>Insights</h1>
<h2>Here are the car accidents that involve more than 3 Vehicles</h2>

<h1>7. Group the accidents by Accident_Severity and count the total number of cases in each category.</h1>

In [56]:
accident.groupby(["Accident_Severity"]).size()

Accident_Severity
Fatal        8661
Serious     88217
Slight     563801
dtype: int64

<h1>Insights</h1>
<h2>We can see now which condition of the accidents has the highest count.</h2>

<h1>8. Group accidents by Road_Surface_Conditions and count the number of accidents in each.</h1>

In [57]:
accident.groupby(["Road_Surface_Conditions"]).size()

Road_Surface_Conditions
Dry                     447821
Flood over 3cm. deep      1017
Frost or ice             18517
Snow                      5890
Wet or damp             186708
unknown condition          726
dtype: int64

<h1>Insights</h1>
<h2>By using groupby, we see each category of Road condition in each accident.</h2>

<h1>9.Correlation between number of casualties and number of vehicles?</h1>

In [58]:
accident["Number_of_Casualties"].corr(accident["Number_of_Vehicles"])

np.float64(0.2288888612692756)

<h1>Insights</h1>
<h2>there is no correlation for Number_of_Casualties and Number_of_Vehicles.</h2>

<h1>10. Which vehicle type is most frequently involved in accidents?
</h1>

In [59]:
accident["Vehicle_Type"].value_counts()

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h1>Insights</h1>

<h2>there was the most frequently involved in accidents is car that has 497992 accidents.</h2>

<h1>Insights for year 2019</h1>

In [60]:
accident[accident['Year'] == 2019]

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
182110       2.01E+12            Slight    2019-12-20  54.985289   
182111       2.01E+12           Serious    2019-12-21  54.984105   
182112       2.01E+12            Slight    2019-12-23  55.166369   
182113       2.01E+12            Slight    2019-12-23  54.995154   
182114       2.01E+12           Serious    2019-12-22  55.104618   

                   Light_Conditions           District Area Longitude  \
0             Darkness - lights lit  Kensington and Chelsea -0.209082   
1                          Daylight  Kensington and Chelsea -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea -0.210215   
3                          Daylight  Kensington and Chelsea -0.202731   
4             Darkness - lights lit  Kensington and Chelsea -0.192487   
...                             ...                     ...       ...   
182110       Darkness - no lighting   Dumfries and Galloway -3.210294   
182111                     Daylight   Dumfries and Galloway -3.193693   
182112       Darkness - no lighting   Dumfries and Galloway -2.992068   
182113        Darkness - lights lit   Dumfries and Galloway -3.058338   
182114                     Daylight   Dumfries and Galloway -3.348857   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
182110                     1                   1            Frost or ice   
182111                     2                   1            Frost or ice   
182112                     1                   1            Frost or ice   
182113                     1                   1             Wet or damp   
182114                     2                   2            Frost or ice   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban        unknown weather   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban        unknown weather   
...                    ...                 ...                    ...   
182110  Single carriageway               Rural                  Other   
182111  Single carriageway               Rural                  Other   
182112  Single carriageway               Rural            Fog or mist   
182113  Single carriageway               Rural     Fine no high winds   
182114  Single carriageway               Rural  Raining no high winds   

                                Vehicle_Type  Year Month DayofWeek Day  
0                                        Car  2019     6         2   5  
1                                        Car  2019     7         1   2  
2                      Taxi/Private hire car  2019     8         0  26  
3       Bus or coach (17 or more pass seats)  2019     8         4  16  
4                              Other vehicle  2019     9         1   3  
...                                      ...   ...   ...       ...  ..  
182110               

In [61]:
year19 = accident[accident['Year'] == 2019]

In [62]:
year19['Accident_Severity'].value_counts()

Accident_Severity
Slight     155079
Serious     24322
Fatal        2714
Name: count, dtype: int64

In [63]:
fatal = year19[year19['Accident_Severity'] =='Fatal']

In [64]:
fatal['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Rural          1773
Urban           941
Unallocated       0
Name: count, dtype: int64

<h1>11. In year 2019, Rural area has 65.32% fatality rate while only 34.66% happened in urban areas indicating that majority of fatal accidents happened in Rural Area</h1>

In [65]:
rural_v_type = fatal[fatal['Urban_or_Rural_Area']=='Rural']

In [66]:
rural_v_type['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      1278
Bus or coach (17 or more pass seats)      105
Van / Goods 3.5 tonnes mgw or under        92
Goods 7.5 tonnes mgw and over              74
Motorcycle over 500cc                      71
Motorcycle 125cc and under                 44
Taxi/Private hire car                      28
Motorcycle over 125cc and up to 500cc      21
Goods over 3.5t. and under 7.5t            17
Motorcycle 50cc and under                  15
Other vehicle                              11
Minibus (8 - 16 passenger seats)            8
Agricultural vehicle                        6
Pedal cycle                                 3
Data missing or out of range                0
Ridden horse                                0
Name: count, dtype: int64

In [67]:
dof_type_rural = rural_v_type.groupby(['DayofWeek','Road_Type']).size()

In [68]:
dof_type_rural.unstack()

Road_Type  Dual carriageway  One way street  Roundabout  Single carriageway  \
DayofWeek                                                                     
0                        52               0           5                 236   
1                        67               3           1                 187   
2                        55               0           1                 180   
3                        45               0           2                 164   
4                        48               0           2                 177   
5                        44               0           3                 191   
6                        49               1           2                 244   

Road_Type  Slip road  
DayofWeek             
0                  3  
1                  0  
2                  2  
3                  4  
4                  2  
5                  1  
6                  2

In [69]:
urban_v_type = fatal[fatal['Urban_or_Rural_Area']=='Urban']

In [70]:
urban_v_type['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      670
Van / Goods 3.5 tonnes mgw or under       48
Bus or coach (17 or more pass seats)      47
Motorcycle over 500cc                     36
Motorcycle 125cc and under                32
Goods 7.5 tonnes mgw and over             32
Taxi/Private hire car                     22
Motorcycle 50cc and under                 20
Motorcycle over 125cc and up to 500cc     16
Goods over 3.5t. and under 7.5t            9
Other vehicle                              4
Minibus (8 - 16 passenger seats)           3
Pedal cycle                                1
Agricultural vehicle                       1
Data missing or out of range               0
Ridden horse                               0
Name: count, dtype: int64

In [71]:
dof_type = urban_v_type.groupby(['DayofWeek','Road_Type']).size()

In [72]:
dof_type.unstack()

Road_Type  Dual carriageway  One way street  Roundabout  Single carriageway  \
DayofWeek                                                                     
0                        23               3           4                  72   
1                        26               3           1                  98   
2                        28               4           1                  97   
3                        25               2           3                  99   
4                        30               8           5                  90   
5                        31               3           0                 109   
6                        34               4           3                 126   

Road_Type  Slip road  
DayofWeek             
0                  1  
1                  1  
2                  2  
3                  0  
4                  3  
5                  1  
6                  1

<h1>12. Cars dominate fatalities in both rural (72.1%) and urban (71.2%) areas.</h1>

<h1>13. Motorcycles have a slightly higher fatality share in urban areas, especially smaller-engine ones (e.g., 50cc and under: 2.1% urban vs. 0.8% rural).</h1>

In [73]:
month_weather19 = fatal.groupby(['Month','Weather_Conditions']).size()

In [74]:
month_weather19.unstack()

Weather_Conditions  Fine + high winds  Fine no high winds  Fog or mist  Other  \
Month                                                                           
1                                  28                 177            0      2   
2                                   4                 128            5      6   
3                                   2                 189            0      2   
4                                   2                 217            2      0   
5                                   6                 194            0      1   
6                                   0                 161            4      4   
7                                   1                 188            0      2   
8                                   0                 224            2      3   
9                                   2                 199            0      3   
10                                  0                 209            8      1   
11                                  9                 179            1     10   
12                                  5                 185            6      9   

Weather_Conditions  Raining + high winds  Raining no high winds  \
Month                                                             
1                                      6                     24   
2                                      3                     24   
3                                      2                     14   
4                                      1                      5   
5                                      4                     24   
6                                      2                     29   
7                                      2                     31   
8                                      1                     28   
9                                      0                     16   
10                                     3                     21   
11                                     8                     28   
12                                    10                     18   

Weather_Conditions  Snowing + high winds  Snowing no high winds  \
Month                                                             
1                                      0                      0   
2                                      0                      1   
3                                      0                      2   
4                                      0                      0   
5                                      0                      0   
6                                      0                      0   
7                                      0                      0   
8                                      0                      0   
9                                      0                      0   
10                                     0                      0   
11                                     0                      0   
12                                     0                      0   

Weather_Conditions  unknown weather  
Month                                
1                                 2  
2                                 2  
3                                 1  
4                                 4  
5                                 1  
6                                 1  
7                                 1  
8                                 2  
9                                 5  
10                                3  
11                                3  
12                                2

<h1>14. Clear weather is the deadliest condition – 76.1% of fatalities happen when the weather is fine with no high winds.</h1>

<h1>15. The months with the most fatalities are August (10.5%) and October (9.9%), indicating potential seasonal risk factors.</h1>

<h1>16. Single carriageways are the deadliest road type in both rural (77.8%) and urban (69.4%) areas.</h1>

<h1>17. Sundays have the highest fatalities in both urban (16.8%) and rural (16.8%) areas.</h1>

<h1>18. Group accidents by Road_Surface_Conditions and count the number of accidents in each.</h1>

In [75]:
accident.groupby(["Road_Surface_Conditions"]).size()

Road_Surface_Conditions
Dry                     447821
Flood over 3cm. deep      1017
Frost or ice             18517
Snow                      5890
Wet or damp             186708
unknown condition          726
dtype: int64

<h1>Insights</h1>
<h2>By using groupby, we see each category of Road condition in each accident.</h2>

<h1>19. Group accidents by Vehicle_Type and find which type has the highest accident severity in each accident condition.</h1>

In [76]:
accident.groupby(["Vehicle_Type", "Accident_Severity"]).size()

Vehicle_Type                           Accident_Severity
Agricultural vehicle                   Fatal                    21
                                       Serious                 282
                                       Slight                 1644
Bus or coach (17 or more pass seats)   Fatal                   325
                                       Serious                3373
                                       Slight                22180
Car                                    Fatal                  6577
                                       Serious               66461
                                       Slight               424954
Data missing or out of range           Fatal                     0
                                       Serious                   0
                                       Slight                    6
Goods 7.5 tonnes mgw and over          Fatal                   216
                                       Serious                2321
     

<h1>Insights</h1>
<h2>The highest Serious accident condition is a Car that has 66461 incidents. The highest Fatal accident condition is a Car that has 6577 incidents. The highest Slight accident condition is a Car that has 424954 incidents.</h2>

<h1>20. Which vehicle type is most frequently involved in accidents?
</h1>

In [77]:
accident["Vehicle_Type"].value_counts()

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h1>Insights</h1>
<h2>there was The most frequently involved in accidents is Car that has 497992 accidents.</h2>